core/offchain-worker-executor/src/executor.rs

```diff
@@ -171,27 +171,24 @@ impl<
 		Ok(())
 	}
 
-	// TODO: this is duplicated code and should be removed, once we refactor the top pool author
-	// and integrate the top pool executor into it.
 	fn remove_calls_from_pool(
 		&self,
 		shard: &ShardIdentifier,
 		executed_calls: Vec<ExecutedOperation>,
 	) -> Vec<ExecutedOperation> {
-		let mut failed_to_remove = Vec::new();
-		for executed_call in executed_calls {
-			if let Err(e) = self.top_pool_author.remove_top(
-				vec![executed_call.trusted_operation_or_hash.clone()],
-				*shard,
-				executed_call.is_success(),
-			) {
-				// We don't want to return here before all calls have been iterated through,
-				// hence only throwing an error log and push to `failed_to_remove` vec.
-				error!("Error removing trusted call from top pool: Error: {:?}", e);
-				failed_to_remove.push(executed_call);
-			}
-		}
-		failed_to_remove
+		let executed_calls_tuple: Vec<_> = executed_calls
+			.iter()
+			.map(|e| (e.trusted_operation_or_hash.clone(), e.is_success()))
+			.collect();
+		let failed_to_remove_hashes =
+			self.top_pool_author.remove_calls_from_pool(*shard, executed_calls_tuple);
+
+		let failed_executed_calls: Vec<_> = executed_calls
+			.into_iter()
+			.filter(|e| failed_to_remove_hashes.contains(&e.trusted_operation_or_hash))
+			.collect();
+
+		failed_executed_calls
```

Comment:
Here we might want to note that we can improve the performance (even I don't know the internals of the code functions here).

* Because `collect` is O(n), and I don't see a strict reason to have `executed_calls_tuple` collected. Could `			self.top_pool_author.remove_calls_from_pool(*shard, executed_calls_tuple);` maybe take an iterator?


I think this function is called very frequently and optimizing here, might really make a difference.

Reply:
Extracted this to a separate issue:  #993